### PR TITLE
[Quant] Add int8 linear op gelu for quantization PT2E with Inductor. input is an int8 CPU tensor; weight is an int8 MdkldnnCPU tensor

### PR DESCRIPTION
**Pull request by Le-Zheng**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116004
* #114854
* #114853
* __->__ #114852

**Summary**
Enable Int8 Linear Gelu post operator fusions for Stock PyTorch Inductor. The input is an int8 CPU tensor and weight is an int8 MkldnnCPU tensor.

**Test plan**
python test/test_quantization.py -k test_qlinear_gelu_pt2e

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by leslie-fang-intel — CHANGES_REQUESTED** [review]
> 2023-11-30T08:01:48.0303273Z /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear.cpp:971:7: error: variable 'post_op_algo' is used uninitialized whenever 'if' condition is false [-Werror,-Wsometimes-uninitialized]
2023-11-30T08:01:48.0305091Z   if (post_op_name == "gelu"){

@Le-Zheng there are some build errors in the preCI, could you help to check it?  Looks like need to initialize `post_op_algo` with default value.

File: test/quantization/core/test_quantized_op.py [hunk 1]
```diff
@@ -4243,6 +4240,23 @@ def test_qlinear_pt2e(self):
...
                 )
 
+    @skipIfNoONEDNN
+    def test_qlinear_linear_relu_pt2e(self):
+        post_op_to_qlinear_ref_dict = {
+            'none': torch.ops.quantized.linear,
+            'relu': torch.ops.quantized.linear_relu,
+        }
+        supported_post_ops = ['none', 'relu']
+        post_op_algorithm = ''
```
Comment (jgong5) [inline test/quantization/core/test_quantized_op.py:4250]: ditto: post_op_algorithms and make it a list.
```suggestion
        post_op_algorithms = []
```

File: test/quantization/core/test_quantized_op.py [hunk 2]
```diff
@@ -4243,6 +4240,23 @@ def test_qlinear_pt2e(self):
...
+        self._test_qlinear_pt2e_helper(post_op_to_qlinear_ref_dict, supported_post_ops, post_op_algorithm)
+
+    @skipIfNoONEDNN
+    def test_qlinear_gelu_pt2e(self):
+        post_op_to_qlinear_ref_dict = {}
+        supported_post_ops = ['gelu']
+        post_op_algorithm = ['none', 'tanh']
```
Comment (jgong5) [inline test/quantization/core/test_quantized_op.py:4257]: ditto

File: test/quantization/core/test_quantized_op.py [hunk 3]
```diff
@@ -4188,9 +4181,9 @@ def test_qlinear_pt2e(self):
         post_op_args = []
         cases = itertools.product(
             in_channels_list, out_channels_list, use_bias_list,
-            supported_post_ops, weight_quant_per_channel_list, output_dtype_list)
+            supported_post_ops, weight_quant_per_channel_list, output_dtype_list, post_op_algorithm)
```
Comment (jgong5) [inline test/quantization/core/test_quantized_op.py:4184]: Make it plural and make it a list.
```suggestion
            supported_post_ops, weight_quant_per_channel_list, output_dtype_list, post_op_algorithms)
```

File: test/quantization/core/test_quantized_op.py (hunk 2 above)
Comment (Le-Zheng) [reply test/quantization/core/test_quantized_op.py]: Thank you for your feedback! I've made the requested changes.

**Comment by Le-Zheng** [comment]
> > 2023-11-30T08:01:48.0303273Z /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear.cpp:971:7: error: variable 'post_op_algo' is used uninitialized whenever 'if' condition is false [-Werror,-Wsometimes-uninitialized]
> > 2023-11-30T08:01:48.0305091Z   if (post_op_name == "gelu"){
> 
> @Le-Zheng there are some build errors in the preCI, could you help to check it? Looks like need to initialize `post_op_algo` with default value.

Thank you for your feedback! I've made the requested changes.

File: test/quantization/core/test_quantized_op.py (hunk 1 above)
Comment (Le-Zheng) [reply test/quantization/core/test_quantized_op.py]: Thank you for pointing out and changed accordingly.

**Comment by Le-Zheng** [comment]
> > 2023-11-30T08:01:48.0303273Z /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qlinear.cpp:971:7: error: variable 'post_op_algo' is used uninitialized whenever 'if' condition is false [-Werror,-Wsometimes-uninitialized]
> > 2023-11-30T08:01:48.0305091Z   if (post_op_name == "gelu"){
> 
> @Le-Zheng there are some build errors in the preCI, could you help to check it? Looks like need to initialize `post_op_algo` with default value.

Thank you for your feedback! I've made the requested changes.

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp [hunk 4]
```diff
@@ -967,7 +967,16 @@ static at::Tensor linear_int8_with_onednn_weight(
   auto bias_desc = with_bias ?
       tensor::desc(onednn_bias.value().get_dims(), ideep::data_type::f32, ideep::format_tag::any) :
       tensor::desc();
-  auto op_attr = onednn_utils::create_attr_by_post_op(post_op_name, post_op_args);
+  dnnl::algorithm post_op_algo = dnnl::algorithm::eltwise_gelu_erf;
```
Comment (leslie-fang-intel) [inline aten/src/ATen/native/quantized/cpu/qlinear.cpp:970]: nit: I think we should initialize the `post_op_algo` with  `dnnl::algorithm::undef`

**Review by leslie-fang-intel — APPROVED** [review]

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp (hunk 4 above)
Comment (Le-Zheng) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp]: Thank you for your comment! I've update accordingly.

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp [hunk 5]
```diff
@@ -967,7 +967,16 @@ static at::Tensor linear_int8_with_onednn_weight(
...
+  if (post_op_name == "gelu"){
+    if (post_op_algorithm == "None"){
```
Comment (leslie-fang-intel) [inline aten/src/ATen/native/quantized/cpu/qlinear.cpp:972]: Should we use `None` or `none`?

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp (hunk 4 above)
Comment (leslie-fang-intel) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp:970]: I think it should be.... Please take a look again.

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp (hunk 5 above)
Comment (Le-Zheng) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp]: Thanks for pointing this typo error and changed accordingly.

**Review by leslie-fang-intel — APPROVED** [review]

**Review by jgong5 — APPROVED** [review]

**Pull request closed without merging**